### PR TITLE
Fixing transport test cases

### PR DESCRIPTION
**Pull request by SudhanAruna**

### CODE REVIEW

File: python/proton/_data.py [hunk 1]
```diff
@@ -705,6 +705,7 @@ def rewind(self):
         the first node.
         """
         assert self._data is not None
+        # print(f"log _data 708: data {self._data}")
```
Comment (jiridanek) [inline python/proton/_data.py:701]: If you have recent enough Python version, you can use `print(f"708: {self._data=}")`. If you add the equal sign, Python will format the print as `708: self._data=the-value`. Cool trick, shows you keep up with Python features ;)

File: python/proton/_wrapper.py [hunk 2]
```diff
@@ -135,5 +141,7 @@ def __repr__(self):
                                         id(self), addressof(self._impl))
 
 
-PYCTX = int(pn_py2void(Wrapper))
+# PYCTX = int(pn_py2void(Wrapper))
+print(f"log _wrapper line 144 : type {type(Wrapper)} \n wrapper {Wrapper}")
+PYCTX = int(Wrapper)
```
Comment (jiridanek) [inline python/proton/_wrapper.py]: My first thought would be to do `PYCTX = ffi.cast('void *', Wrapper)` here. I'll have a closer look in the evening and try it out.

The intent of this code seems to me to keep around the pointer and for some reason it is represented on the Python side as an int. No reason for that, IMO. Keep it be a void pointer.

Comment (jiridanek) [reply python/proton/_wrapper.py]: Uh, disregard that. Hmm, this is in the realm of callbacks and things like that. https://cffi.readthedocs.io/en/latest/using.html#extern-python-new-style-callbacks I'll need to understand the use and look into this more; figure out how to accomplish this with CFFI.

File: python/proton/_wrapper.py [hunk 3]
```diff
@@ -134,6 +150,12 @@ def __repr__(self):
                                         self.__class__.__name__,
                                         id(self), addressof(self._impl))
 
+# print(f"log _wrapper line 144 {pn_py2void(Wrapper)=} ")
+# PYCTX = int(pn_py2void(Wrapper))
+# print(f"log _wrapper line 144 : type {type(Wrapper)} \n wrapper {Wrapper}")
+handle = ffi.new_handle(Wrapper)
```
Comment (SudhanAruna) [inline python/proton/_wrapper.py:147]: @jiridanek I'm sure about this approach. Using ffi.new_handle() has return random unique value. This didn't raise any errors here

**Comment by SudhanAruna** [comment]
FYI I'm debugging the issues so I have left the print functions

File: python/proton/_data.py (hunk 1 above)
Comment (SudhanAruna) [reply python/proton/_data.py:701]: That's very useful trick

File: python/proton/_wrapper.py [hunk 4]
```diff
@@ -78,11 +83,12 @@ def __init__(self, impl_or_constructor, get_context=None):
 
         if get_context:
             record = get_context(impl)
-            attrs = pn_void2py(pn_record_get(record, PYCTX))
-            if attrs is None:
+            attrs = pn_record_get(record, PYCTX)
+            if attrs is None or attrs == ffi.NULL:
                 attrs = {}
+                attrs = ffi.new_handle({})
                 pn_record_def(record, PYCTX, PN_PYREF)
```
Comment (SudhanAruna) [inline python/proton/_wrapper.py:90]: @jiridanek  The code is crashing in this line. After executing `pn_record_def` and `pn_record_set` it crashes without any useful error message. Remaining test cases also depend on the _wrapper.py.
```
>>> attrs
<cdata 'void *' handle to {}>
>>> pn_record_set(record, pyctx, attrs)
>>> pn_record_def(record, pyctx, PN_PYREF)
>>> pn_record_set(record, pyctx, attrs)
`Segmentation fault (core dumped)`
````

**Comment by jiridanek** [comment]
<jdanek> SudhanAruna, regarding debug  prints, did you mention that you did not see print output when you were debugging the Wrapper issues?
<jdanek> what seems to work is to throw exceptions, instead of prints
<jdanek> raise RuntimeError("debug message")

**Comment by jiridanek** [comment]
I've managed to bring it to the following error

```
28: ======================================================================
28: ERROR: testEOS (test_transport.ClientTransportTest)
28: ----------------------------------------------------------------------
28: Traceback (most recent call last):
28:   File "/home/jdanek/repos/qpid/qpid-proton/python/tests/proton_tests/test_transport.py", line 39, in setUp
28:     self.conn = Connection()
28:   File "/home/jdanek/repos/qpid/qpid-proton/python/proton/_endpoints.py", line 169, in __init__
28:     Wrapper.__init__(self, impl, pn_connection_attachments)
28:   File "/home/jdanek/repos/qpid/qpid-proton/python/proton/_wrapper.py", line 86, in __init__
28:     attrs = pn_record_get(record, PYCTX)
28: TypeError: initializer for ctype 'struct pn_record_t *' must be a cdata pointer, not SwigPyObject
28: 
28: ----------------------------------------------------------------------
28: Ran 111 tests in 0.058s
28: 
28: FAILED (errors=1, skipped=28)
1/1 Test #28: python-test ......................***Failed  Required regular expression not found. Regex=[Ran .+ tests in .*
```

so at this point it needs your updated _endpoints.py from the other file (because now SWIG object is leaking in).

I chose to believe I am on the right path and I will merge what I have. Your work with the CFFI new_handler function was good discovery, that is probably the correct way to deal with this. The crash was happening because you did not initialize PN_PYREF, which is what I am now doing.

**Pull request merged**